cache/cache.go

```diff
@@ -250,24 +256,61 @@ func (c *Cache) watch(ctx context.Context, rev int64) error {
 			clientv3.WithCreatedNotify(),
 		)
 
-		for resp := range watchCh {
-			readyOnce.Do(func() { c.ready.Set() })
-			if err := resp.Err(); err != nil {
-				c.ready.Reset()
-				c.demux.Purge()
-				return err
+		err := c.watchEvents(ctx, watchCh, applyErr, &readyOnce)
+		c.demux.Unregister(storeW)
+
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (c *Cache) applyStorage(ctx context.Context, storeW *watcher, applyErr chan<- error) {
+	defer c.waitGroup.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.internalCtx.Done():
+			return
+		case events, ok := <-storeW.eventQueue:
+			if !ok {
+				return
+			}
+			if err := c.store.Apply(events); err != nil {
+				select {
+				case applyErr <- err:
+				default:
+				}
```

Comment:
```suggestion
				applyErr <- err
```
I don't think you should skip error. Even if this should not happen it's bad to imply that sending error is optional.